### PR TITLE
WL-3900 ZIP and CC from Resources both download to the desktop,

### PR DESCRIPTION
**Pull request by RebeccaMiller-Which**
incorporates SAK-41272 Code improvements for Zip download in Resources. (#6552)

### CODE REVIEW

File: velocity/tool/src/java/org/sakaiproject/cheftool/VelocityPortletPaneledAction.java [hunk 1]
```diff
@@ -205,7 +205,7 @@ public static void addAlert(SessionState state, String message)
 		String soFar = (String) state.getAttribute(STATE_MESSAGE);
 		if (soFar != null)
 		{
-			soFar = soFar + "\n\n" + message;
+			soFar = soFar + "<br/>" + message;
```
Comment (buckett) [inline velocity/tool/src/java/org/sakaiproject/cheftool/VelocityPortletPaneledAction.java]: Is this ok? This is used outside of resources.

File: content/content-tool/tool/src/java/org/sakaiproject/content/exception/ZipMaxTotalSizeException.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package org.sakaiproject.content.exception;
+
+import org.sakaiproject.exception.SakaiException;
+
+/**
+ * <p>
+ * ZipMaxTotalSizeException is thrown whenever the size of all files to be zipped exceeds the maximum permitted in default.sakai.properties.
+ * </p>
+ */
+public class ZipMaxTotalSizeException extends SakaiException {
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/exception/ZipMaxTotalSizeException.java:10]: Comment for future (don't bother changing): I would have been tempted to embed the size of the limit in the exception. It then makes it easier when displaying a message to the user to be able to say what the limit it.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 3]
```diff
@@ -10521,5 +10397,229 @@ public void doViewTrash(RunData data)
...
+			ContentEntity myElement = (ContentEntity) it.next();
+			if (myElement.isResource()) 
+			{
+				long tempSize = ((ContentResource)myElement).getContentLength();
+				if (tempSize > maxIndividualFileSize) {
+					// Work out the file path without the site ID.
+					String filePath = myElement.getId().replace("/group/" + ToolManager.getCurrentPlacement().getContext(), "");
+					zipSingleFileSizeExceeded.add(filePath);
+				}
+				else {total=total+tempSize;}
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: Sakai normally splits this out onto multiple lines.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 4]
```diff
@@ -10521,5 +10397,229 @@ public void doViewTrash(RunData data)
...
+			}
+			else if (myElement.isCollection())
+			{
+				long tempSize = getCollectionRecursiveSize((ContentCollection)myElement, maxIndividualFileSize, zipMaxTotalSize, zipSingleFileSizeExceeded);
+				if (tempSize > zipMaxTotalSize) {
+					throw new ZipMaxTotalSizeException();
+				}
+				else {total=total+tempSize;}
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: Sakai normally splits this out onto multiple lines.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java (hunk 3 above)
Comment (buckett) [reply content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: If you changes it you can use `total += tempSize;`

File: content/content-tool/tool/src/webapp/vm/content/sakai_resources_list.vm [hunk 5]
```diff
@@ -80,7 +80,7 @@
 		</ul>
 		</ul>
 	#if ($alertMessage)
-                 <br /><div class="messageValidation">$alertMessage</div><div class="clear"></div>
+                 <br /><div class="alertMessage">$alertMessage</div><div class="clear"></div>
```
Comment (buckett) [inline content/content-tool/tool/src/webapp/vm/content/sakai_resources_list.vm:83]: This seems like the right change to make, did you notice any other code using this block?

**Review by buckett — APPROVED** [review]
Is some of this a cherry-pick?

File: kernel/kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java [hunk 6]
```diff
@@ -82,6 +84,59 @@ public static int getMaxZipExtractFiles() {
...
+			// shouldn't happen by this stage.
+			LOG.error(e.getMessage(), e);
+		}
+
+		try (OutputStream zipOut = response.getOutputStream(); ZipOutputStream out = new ZipOutputStream(zipOut)) {
+			// If in dropbox need to add the word Dropbox to the end of the zip filename - use the first entry in the resourcesToZip map to find if we are in the dropthe user ID.
+			if (!resourcesToZip.isEmpty()) {
+				String firstContentResourceId = resourcesToZip.entrySet().iterator().next().getKey();
+				if (ContentHostingService.isInDropbox(firstContentResourceId) && ServerConfigurationService.getBoolean("dropbox.zip.haveDisplayname", true)) {
+					response.setHeader("Content-disposition", "inline; filename=" + siteId + "DropBox.zip");
```
Comment (buckett) [inline kernel/kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java]: If you're supplying a filename then the disposition should be `attachment` I think.

File: velocity/tool/src/java/org/sakaiproject/cheftool/VelocityPortletPaneledAction.java (hunk 1 above)
Comment (RebeccaMiller-Which) [reply velocity/tool/src/java/org/sakaiproject/cheftool/VelocityPortletPaneledAction.java]: I'll shift it to the types.properties file, there is other HTML in there.

File: content/content-tool/tool/src/java/org/sakaiproject/content/exception/ZipMaxTotalSizeException.java (hunk 2 above)
Comment (RebeccaMiller-Which) [reply content/content-tool/tool/src/java/org/sakaiproject/content/exception/ZipMaxTotalSizeException.java:10]: Good point.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java (hunk 3 above)
Comment (RebeccaMiller-Which) [reply content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: Have changed this.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java (hunk 4 above)
Comment (RebeccaMiller-Which) [reply content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: Have changed this.

File: kernel/kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java (hunk 6 above)
Comment (RebeccaMiller-Which) [reply kernel/kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java]: Ok thanks.

File: content/content-tool/tool/src/webapp/vm/content/sakai_resources_list.vm (hunk 5 above)
Comment (RebeccaMiller-Which) [reply content/content-tool/tool/src/webapp/vm/content/sakai_resources_list.vm:83]: All the other velocity pages that use the alertMessage in the context have the class as alertMessage.

**Comment by RebeccaMiller-Which** [comment]
Changes made.  This is a cherry-pick from Sakai but with the export to common cartridge functionality that was already in WL added.

**Pull request merged**